frame/contracts/src/lib.rs

```diff
@@ -1035,40 +1101,55 @@ impl<T: Trait + Send + Sync> SignedExtension for CheckBlockGasLimit<T> {
 	type Call = <T as Trait>::Call;
 	type AdditionalSigned = ();
 	type DispatchInfo = DispatchInfo;
-	type Pre = ();
+	/// Optional pre-dispatch check data.
+	///
+	/// It is present only for the contract calls that operate with gas.
+	type Pre = Option<DynamicWeightData<T::AccountId, NegativeImbalanceOf<T>>>;
 
-	fn additional_signed(&self) -> sp_std::result::Result<(), TransactionValidityError> { Ok(()) }
+	fn additional_signed(&self) -> Result<(), TransactionValidityError> { Ok(()) }
+
+	fn pre_dispatch(
+		self,
+		who: &Self::AccountId,
+		call: &Self::Call,
+		_: DispatchInfo,
+		_: usize,
+	) -> Result<Self::Pre, TransactionValidityError> {
+		Self::perform_pre_dispatch_checks(who, call)
+			.map_err(Into::into)
+	}
 
 	fn validate(
 		&self,
-		_: &Self::AccountId,
+		who: &Self::AccountId,
 		call: &Self::Call,
 		_: Self::DispatchInfo,
 		_: usize,
 	) -> TransactionValidity {
-		let call = match call.is_sub_type() {
-			Some(call) => call,
-			None => return Ok(ValidTransaction::default()),
-		};
+		Self::perform_pre_dispatch_checks(who, call)
```

Comment:
Hmm, wouldn't this do the actual withdrawal instead of just checking `ensure_can_withdraw`?

Reply:
It will, but I assumed that's alright since we are in validation context therefore all changes are discarded. Or you see some other problem?